### DISCUSSION TITLE
best strategy to stream play an MP3 from url while saving it to SD card

### DISCUSSION THREAD

**Discussion by 0x0fe** [pschatzmann/arduino-audio-tools#507] [Q&A] [answered]
So, assuing a remote MP3 file such as : https://file-examples.com/storage/fe352586866388d59a8918d/2017/11/file_example_MP3_1MG.mp3

what would be the best strategy to save it to the SD card, while playing it. 
The idea being to play the file almost immediately without waiting for the download to complete, we could just wait enough to have a little bit of buffering.

I havent found such option in audio-tools but i may have missed it since the library is quite extensive.
Thanks.

**Comment by pschatzmann** [comment, marked as answer]
Just copy the data from an URLStream to a MultiOutput which will go to a File and to an EncodedAudioStream.

**Reply by 0x0fe** [reply]
oh, interesting, thank you for the hint, i will try that

**Reply by pschatzmann** [reply]
The second option is to do the copy loop yourself: in a loop, just read from the URLStream (with readBytes()) and then write the data both to the EncodedAudioStream and your file...

**Comment by 0x0fe** [comment]
while trying this i realized i dont have the file length, is there a way to get the file length (when this is a file and not an endless stream) from audiocopy or another function? for example i noticed i got [W] HttpRequest.h : 153 - no CONTENT_LENGTH found in reply
when opening stream, would it mean i can get CONTENT_LENGTH when it is an actual file?

**Comment by pschatzmann** [comment]
Sure it is in the HttpReplyHeader where you can just call get(CONTENT_LENGTH) on the attribute name CONTENT_LENGTH which returns a string. or use getReceivedContentLength() on the HttpRequest if you need a integer.
You best use the API documentation to figure out how to access this.
